### DISCUSSION TITLE
17. HTTPS와 SSL Handshake에 대해 설명해 주세요.

### DISCUSSION THREAD

**Discussion by cutehumanS2** [jewelry-box/cs-interview#4] [General]
HTTPS와 SSL Handshake에 대해 설명해 주세요.

**Comment by jongmee** [comment]
- HTTPS는 HTTP에서 `SSL/TLS`로 데이터를 `암호화`해서 보안을 강화한 프로토콜이다. 데이터를 암호화해서 도청이나 중간자 공격을 방지한다. 또한 전송된 데이터가 도중에 변조되지 않았음을 보장하고 웹사이트가 신뢰할 수 있는 기관에 의해 인증되었음을 보장한다.

- 연결할 때 TCP 3 way handshake 이후에 `SSL handshake`를 진행한다.

- SSL handshake 진행 과정에서 `대칭키` 암호화 방식과 `비대칭키` 암호화 방식을 모두 사용한다.
  1. 클라이언트가 서버에게 임의의 난수와 사용할 수 있는 암호화 방식과 함께 `인증서를 요청`한다.
  2. 서버는 `서버의 공개키`가 담긴 SSL 인증서와 임의의 난수를 응답한다. 이 인증서는 `CA의 비밀키`로 암호화되어 있다.
  3. 클라이언트는 `CA의 공개키`로 인증서를 해독해서 신뢰도 있는 CA에서 발급한 인증서인지를 확인하고 난수를 활용해서 시크릿을 만들고 `서버의 공개키`로 암호화해서 서버에 전송한다.
  4. 서버는 `서버의 비밀키`로 시크릿을 복호화한 후 추후 통신에서 사용할 `대칭키`를 만들어서 교환한다.
